cli/cmd/generate_gcp.go

```diff
@@ -0,0 +1,780 @@
+		// validation to know which option was selected due to how survey works; and doing it by index (also supported) is
+		// difficult when the options are dynamic (which they are)
+		options := []string{GcpAdvancedOptAuditLog, GcpAdvancedOptExistingServiceAccount, GcpAdvancedOptIntegrationName}
+
+		options = append(options, GcpAdvancedOptLocation, GcpAdvancedOptDone)
+		if err := SurveyQuestionInteractiveOnly(SurveyQuestionWithValidationArgs{
+			Prompt: &survey.Select{
+				Message: "Which options would you like to configure?",
+				Options: options,
+			},
+			Response: &answer,
+		}); err != nil {
+			return err
+		}
+
+		// Based on response, prompt for actions
+		switch answer {
+		case GcpAdvancedOptAuditLog:
+			if err := promptGcpAuditLogQuestions(config, extraState); err != nil {
+				return err
+			}
+		case GcpAdvancedOptExistingServiceAccount:
+			if err := promptGcpExistingServiceAccountQuestions(config); err != nil {
+				return err
+			}
+		case GcpAdvancedOptIntegrationName:
+			if err := promptGcpIntegrationNameQuestions(config); err != nil {
+				return err
+			}
+		case GcpAdvancedOptLocation:
+			if err := promptCustomizeGcpOutputLocation(extraState); err != nil {
+				return err
+			}
+		}
+
+		// Re-prompt if not done
+		innerAskAgain := true
+		if answer == GcpAdvancedOptDone {
+			innerAskAgain = false
+		}
+
+		if err := SurveyQuestionInteractiveOnly(SurveyQuestionWithValidationArgs{
+			Checks:   []*bool{&innerAskAgain},
+			Prompt:   &survey.Confirm{Message: QuestionGcpAnotherAdvancedOpt, Default: false},
+			Response: &innerAskAgain,
+		}); err != nil {
+			return err
+		}
+
+		if !innerAskAgain {
+			answer = GcpAdvancedOptDone
+		}
+	}
+
+	return nil
+}
+
+func configOrAuditLogEnabled(config *gcp.GenerateGcpTfConfigurationArgs) *bool {
+	auditLogOrConfigEnabled := config.AuditLog || config.Config
+	return &auditLogOrConfigEnabled
+}
+
+func gcpConfigIsEmpty(g *gcp.GenerateGcpTfConfigurationArgs) bool {
+	return !g.AuditLog &&
+		!g.Config &&
+		g.ServiceAccountCredentials == "" &&
+		g.GcpOrganizationId == "" &&
+		g.LaceworkProfile == ""
+}
+
+func writeGcpGenerationArgsCache(a *gcp.GenerateGcpTfConfigurationArgs) {
+	if !gcpConfigIsEmpty(a) {
+		// If ExistingIamRole is partially set, don't write this to cache; the values won't work when loaded
+		if a.ExistingServiceAccount.IsPartial() {
+			a.ExistingServiceAccount = nil
+		}
+		cli.WriteAssetToCache(CachedGcpAssetIacParams, time.Now().Add(time.Hour*1), a)
+	}
+}
+
+// entry point for launching a survey to build out the required generation parameters
+func promptGcpGenerate(
+	config *gcp.GenerateGcpTfConfigurationArgs,
+	existingServiceAccount *gcp.ExistingServiceAccountDetails,
+	extraState *GcpGenerateCommandExtraState,
+) error {
+	// Cache for later use if generation is abandon and in interactive mode
+	if cli.InteractiveMode() {
+		defer writeGcpGenerationArgsCache(config)
+		defer extraState.writeCache()
```

Comment:
Why these two state structures have different cache functions, this one in embedded and the other one above not?

Reply:
It's because there is a need to do further validation on the generation args. Inside `writeGcpGenerationArgsCache` there's a check/validation that happens to ensure all required structs are populated enough to actually cache (so we don't end up in weird states when resuming).  

(https://github.com/lacework/go-sdk/blob/c11b799441b5e641984e12781f292e9f8281640d/cli/cmd/generate_gcp.go#L688)